From c9b7b02406c7965d4308763c0bb8d702bcfbba18 Mon Sep 17 00:00:00 2001
From: Andre Amegah <andrex.amegah@intel.com>
Date: Tue, 30 Apr 2013 17:47:17 +0300
Subject: [PORT FROM MAIN] Telephony: Support for network seach mode change event

BZ: 63416

The STK Network Search Mode Change event is not supported
in Android framework.

This patch provides support for Network Search Mode Change
event.

Network Search Mode Change event is an USAT Event Download
command.

If an UICC Sat Application registers to this event through
SetUpEventList command and then, when the terminal detects
a change in current Network search mode (either manual
or automatic) the terminal will inform UICC Application
using an Envelope command.

Change-Id: I23048b011f6da1e634446c5546b17d2a6700ae55
Orig-Change-Id: I5f3170990d549bd287f729b588c252db0c443d14
Signed-off-by: Andre Amegah <andrex.amegah@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 .../android/internal/telephony/cat/CatService.java |   37 ++++++++++++++++++++
 1 files changed, 37 insertions(+), 0 deletions(-)

diff --git a/src/java/com/android/internal/telephony/cat/CatService.java b/src/java/com/android/internal/telephony/cat/CatService.java
index 5186c87..d4718b2 100644
--- a/src/java/com/android/internal/telephony/cat/CatService.java
+++ b/src/java/com/android/internal/telephony/cat/CatService.java
@@ -29,6 +29,7 @@ import android.os.HandlerThread;
 import android.os.Message;
 import android.os.RemoteException;
 import android.os.SystemProperties;
+import android.telephony.ServiceState;
 
 import com.android.internal.telephony.CommandsInterface;
 import com.android.internal.telephony.Phone;
@@ -99,6 +100,7 @@ public class CatService extends Handler implements AppInterface {
     static final int MSG_ID_SIM_READY                = 7;
 
     static final int MSG_ID_RIL_MSG_DECODED          = 10;
+    static final int MSG_ID_NETWORK_SEARCH_MODE_CHANGE   = 11;
 
     // Events to signal SIM presence or absent in the device.
     private static final int MSG_ID_ICC_RECORDS_LOADED       = 20;
@@ -114,6 +116,8 @@ public class CatService extends Handler implements AppInterface {
     static final int NETWORK_SELECTION_MODE_MANUAL         = 0x00;
     static final int NETWORK_SELECTION_MODE_AUTOMATIC      = 0x01;
 
+    private int mCurrentNetworkSelectionMode;
+
     static final String STK_DEFAULT = "Default Message";
 
     /* Intentionally private for singleton */
@@ -158,6 +162,7 @@ public class CatService extends Handler implements AppInterface {
         mCmdIf.unSetOnCatCallSetUp(this);
 
         removeCallbacksAndMessages(null);
+        PhoneFactory.getDefaultPhone().unregisterForServiceStateChanged(this);
     }
 
     @Override
@@ -369,6 +374,11 @@ public class CatService extends Handler implements AppInterface {
                     for (byte b : mEventList) {
                         CatLog.d(this, "Registered Event: " + b);
                     }
+                    if (isEventDownloadActive(EventCode.NETWORK_SEARCH_MODE_CHANGE.value())) {
+                        mCurrentNetworkSelectionMode = getNetworkSelectionMode();
+                        PhoneFactory.getDefaultPhone().registerForServiceStateChanged(
+                                this, MSG_ID_NETWORK_SEARCH_MODE_CHANGE, null);
+                    }
                 } else {
                     CatLog.d( this, "WARNING: No Event in event list!" );
                 }
@@ -751,6 +761,9 @@ public class CatService extends Handler implements AppInterface {
             CatLog.d(this, "SIM ready. Reporting STK service running now...");
             mCmdIf.reportStkServiceIsRunning(null);
             break;
+        case MSG_ID_NETWORK_SEARCH_MODE_CHANGE:
+            onNetworkSearchModeChange(msg);
+            break;
         default:
             throw new AssertionError("Unrecognized CAT command: " + msg.what);
         }
@@ -917,4 +930,28 @@ public class CatService extends Handler implements AppInterface {
 
         return isManual ? NETWORK_SELECTION_MODE_MANUAL : NETWORK_SELECTION_MODE_AUTOMATIC;
     }
+
+    private void onNetworkSearchModeChange(Message msg) {
+        if (isEventDownloadActive(EventCode.NETWORK_SEARCH_MODE_CHANGE.value())) {
+            ServiceState state = (ServiceState)((AsyncResult) msg.obj).result;
+
+            boolean isManual = state.getIsManualSelection();
+            int mode = isManual ? NETWORK_SELECTION_MODE_MANUAL : NETWORK_SELECTION_MODE_AUTOMATIC;
+
+            if ((mCurrentNetworkSelectionMode ^ mode) != 0) {
+                mCurrentNetworkSelectionMode = mode;
+                byte[] additionalInfo = new byte[3];
+                additionalInfo[0] = (byte)ComprehensionTlvTag.NETWORK_SEARCH_MODE.value();
+                additionalInfo[1] = 0x01;
+                additionalInfo[2] = (byte)mode;
+
+                CatLog.d(this, "Network selection mode is " + (isManual ? "MANUAL" : "AUTOMATIC"));
+
+                onEventDownload(new CatEventMessage(
+                        EventCode.NETWORK_SEARCH_MODE_CHANGE.value(), additionalInfo, false));
+            }
+        } else {
+            PhoneFactory.getDefaultPhone().unregisterForServiceStateChanged(this);
+        }
+    }
 }
-- 
1.7.4.1

